controllers/kuadrant_status.go

```diff
@@ -0,0 +1,85 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	meta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	kuadrantv1beta1 "github.com/kuadrant/kuadrant-operator/api/v1beta1"
+	"github.com/kuadrant/kuadrant-operator/pkg/common"
+)
+
+const (
+	ReadyConditionType string = "Ready"
+)
+
+func (r *KuadrantReconciler) reconcileStatus(ctx context.Context, kObj *kuadrantv1beta1.Kuadrant, specErr error) (ctrl.Result, error) {
+	logger := logr.FromContext(ctx)
+	newStatus := r.calculateStatus(kObj, specErr)
+
+	equalStatus := kObj.Status.Equals(newStatus, logger)
```

Comment:
I suppose this is to prevent unnecessary updates on the status, where a minimum spottable diff would be the value of `LastTransitionTime` of a particular condition, is that right?

Is it worth the cost of checking while:
1. leader election is enabled and therefore only one replica at a time will reconcile the resource?
2. status update is an idempotent operation?

Please, no need to change anything. I'm just trying to understand the flow.

Reply:
To prevent unnecessary updates on the status, specifically related to the `LastTransitionTime` field, conditions are first copied from the existing CR and then desired values applied. After that, they are compared. The status is only updated when there are changes. 

> leader election is enabled and therefore only one replica at a time will reconcile the resource?

I do not fully understand this point. The controller does not expect anyone to change the status. But even if it changes, the controller will set its desired state.

> status update is an idempotent operation?

Yes it is. Identical status requests should always turn into a identical status objects in the CR.

Reply:
> leader election is enabled and therefore only one replica at a time will reconcile the resource?
>> I do not fully understand this point. The controller does not expect anyone to change the status. But even if it changes, the controller will set its desired state.

When leader election is enabled in a manager, no matter how many replicas of the operator are running, only the leader will perform the reconciliation tasks under this manager; therefore, also the status update. IOW, I think there's no concurrency to worry about. (Unless reconcilers are invoked in concurrent goroutines. Are they? 🤔)

> status update is an idempotent operation?
>> Yes it is.

So why carrying checking for changes in the subresource? Because sorting the array and marshalling 2 JSON objects cost less than sending the update request? If the update happens in the background (because we don't have to wait until it finishes to return from reconciling, right?), wouldn't it be better to just trust idempotence and skip the check?

Reply:
In [Authorino](https://github.com/Kuadrant/authorino/blob/20ed001f50544ec25e00b9a62e15073b46798274/controllers/auth_config_status_updater.go#L74) and in [Authorino Operator](https://github.com/Kuadrant/authorino-operator/blob/2d97c6044b82c7b70e850404e382bba6ebf60cfa/controllers/authorino_controller.go#L739) we went with a different approach, where we trust idempotence of the status update operation and leave the rest to the API server.

But it's your call.

> one update request has much more cost (I would say several orders of magnitude) than serializing in memory struct objects which are not expected to be "big"

Maybe I wasn't clear before. You don't have to wait for the state update to finish. You can fork a goroutine to handle the update and return immediately. So, from the perspective of freeing the reconciler/closing the connection, performing the update without the check is theoretically faster.

On the other hand, I totally understand if you don't want to handle this in the background (it could fail!)

And, as you said, sorting two arrays of 5-6 elements each + serializing two small structs as JSON, we're talking nanoseconds might not worth saving.

Anyway, the PR LGTM. As I said before, I was just trying to understand the flow. Not suggesting changing anything.